### PR TITLE
JSR375: support overriding the login configuration by global client certificate.

### LINKED ISSUE #3162: JSR375: support overriding the login configuration by global client certificate.
support overrideHttpAuthenticationMechanism=CLIENT_CERT option which overrides login configuration for both the legacy application and the JSR375 application with the client certificate authentication.
For the the legacy application, failover setting is unchanged. For JSR375 application, if allowAuthenticationFailOverToAuthMethod is set to FORM or BASIC, or allowFailOverToBasicAuth is set to true, it'll fall back to the original login configuration when the client cert login fails.

### PR DESCRIPTION
**Pull request by toshiyamamoto**
fixes #3162

### CODE REVIEW

**Comment by toshiyamamoto** [comment]
#build

File: dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties [hunk 1]
```diff
@@ -91,8 +91,7 @@ trackLoggedOutSSOCookies=Track logged out LTPA single signon tokens
 trackLoggedOutSSOCookies.desc=Specifies whether to track LTPA single signon tokens that are logged out on a server so that it can not be reused on the same server.
 
 overrideHttpAuthenticationMechanism=Name of HttpAuthenticationMechanism to override
-overrideHttpAuthenticationMechanism.desc=Specifies the HttpAuthenticationMechanism for overriding an application-defined HttpAuthenticationMechanism. The Java EE 8 Security specification defines the HttpAuthenticationMechanism. The acceptable value is BASIC or FORM. When FORM is used, specify the loginFormURL and loginErrorURL properties.
-
+overrideHttpAuthenticationMechanism.desc=Specifies the login configuration for overriding an application defined login configuration. The acceptable value is BASIC, FORM or CLIENT_CERT. When BASIC or FORM is used, it is specifies the HttpAuthenticationMechanism, of which the JAVA EE 8 security specification defines, for overriding an application-defined HttpAuthenticationMechanism. When CLIENT_CERT is used, the client certificate authentication takes precedence for the original login configuration of all applications.
```
Comment (chirp1) [inline dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties]: Change: When BASIC or FORM is used, it is specifies the HttpAuthenticationMechanism, of which the JAVA EE 8 security specification defines, for overriding an application-defined HttpAuthenticationMechanism. When CLIENT_CERT is used, the client certificate authentication takes precedence for the original login configuration of all applications.
To:
The BASIC or FORM value specifies the HttpAuthenticationMechanism bean for overriding an application-defined HttpAuthenticationMechanism bean. When you use the CLIENT_CERT value, the client certificate authentication takes precedence for the original login configuration of all applications.

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

File: dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties (hunk 1 above)
Comment (toshiyamamoto) [reply dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties]: Thank you for your review, unfortunately, the message has been revised after your review. Would you mind reviewing the latest one again?

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

File: dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties [hunk 2]
```diff
@@ -76,28 +76,28 @@ useAuthenticationDataForUnprotectedResource=Use authenticated identity for an un
...
-loginErrorURL.desc=Specifies the global URL of a form login error page that includes the root context. The form login error page must be part of a WAR file. An application uses the global login error URL if the application uses form login authentication and does not specify either the form login page or the login error page in the auth-method element of the web.xml file.
+loginErrorURL.desc=Specifies the global URL of a form login error page that includes the root context. The form login error page must be part of a WAR file. An application uses the global login error URL if the application uses form login authentication and does not specify either the form login page or the login error page in the auth-method element of the web.xml file. This is required when overrideHttpAuthMethod attribute is set to FORM.
 
-allowAuthenticationFailOverToAuthMethod=Allow certificate authentication fail over
-allowAuthenticationFailOverToAuthMethod.desc= Specifies the authentication fail over method that will be used when the certificate authentication fails. Valid values are BASIC and FORM.
+allowAuthenticationFailOverToAuthMethod=Fail over method when certificate login fails
```
Comment (bsbyrd1) [inline dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties]: When used as an adjective, Failover is one word. Change label to:
Failover method when certificate login fails

File: dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties [hunk 3]
```diff
@@ -76,28 +76,28 @@ useAuthenticationDataForUnprotectedResource=Use authenticated identity for an un
...
+overrideHttpAuthMethod.desc=Specifies the authentication method to be used for all applications. This specified value overrides any application defined authentication method. The acceptable value is BASIC, FORM or CLIENT_CERT. When FORM is used, loginFormURL and loginErrorURL attributes are required to be set.
 
 contextRootForFormAuthenticationMechanism=Context root for the Java EE 8 login form
-contextRootForFormAuthenticationMechanism.desc=Specifies the context root of a form login page, which is specified by the loginFormURL property. If this value is not set, the first path element of the loginFormURL property is used as a context root. This value is applicable when overrideHttpAuthenticationMechanism is set to FORM.
+contextRootForFormAuthenticationMechanism.desc=Specifies the context root of a form login page, which is specified by the loginFormURL property. If this value is not set, the first path element of the loginFormURL property is used as a context root. This value is applicable when overrideHttpAuthMethod is set to FORM, or if overrideHttpAuthMethodis set to CLIENT_CERT and allowAuthenticationFailOverToAuthMethod is set to FORM.
```
Comment (bsbyrd1) [inline dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties]: Change: 
if overrideHttpAuthMethodis set to CLIENT_CERT
To:
if overrideHttpAuthMethodis is set to CLIENT_CERT

Comment (toshiyamamoto) [reply dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties]: Fixed. Thank you.

File: dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties (hunk 2 above)
Comment (toshiyamamoto) [reply dev/com.ibm.ws.webcontainer.security.app/resources/OSGI-INF/l10n/metatype.properties]: Fixed.

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Comment by toshiyamamoto** [comment]
#build

**Pull request merged**